extensions/amp-fit-text/1.0/fit-text.js

```diff
@@ -0,0 +1,152 @@
+
+  const resize = useCallback(
+    (maxHeight, maxWidth) => {
+      if (!measurerRef.current) {
+        return;
+      }
+      const fontSize = calculateFontSize(
+        measurerRef.current,
+        Number(maxHeight),
+        Number(maxWidth),
+        minFontSize,
+        maxFontSize
+      );
+      getOverflowStyle(measurerRef.current, Number(maxHeight), fontSize);
+    },
+    [maxFontSize, minFontSize]
+  );
+
+  // Here and below, useLayoutEffect is used so intermediary font sizes
+  // during resizing are resolved before the component visually updates.
+  // Font size should readjust when container resizes.
+  useLayoutEffect(() => {
+    const node = contentRef.current;
+    if (!node) {
+      return;
+    }
+    const observer = new ResizeObserver((entries) => {
+      const last = entries[entries.length - 1];
+      const {height: maxHeight, width: maxWidth} = last['contentRect'];
+      resize(maxHeight, maxWidth);
+    });
+    observer.observe(node);
+    return () => observer.disconnect();
+  }, [resize]);
+
+  // Font size should readjust when content changes.
+  useLayoutEffect(() => {
+    resize(height, width);
+  }, [children, resize, height, width]);
+
+  return (
+    <div {...rest}>
+      <div
+        ref={contentRef}
+        style={{
+          ...styles.fitTextContent,
+          'width': px(width),
+          'height': px(height),
+        }}
+      >
+        <div
+          ref={measurerRef}
+          style={{
+            ...styles.fitTextContentWrapper,
+          }}
+        >
+          {children}
+        </div>
+      </div>
+    </div>
+  );
+}
+
+/**
+ * @param {Element} measurer
+ * @param {number} expectedHeight
+ * @param {number} expectedWidth
+ * @param {number} minFontSize
+ * @param {number} maxFontSize
+ * @return {number}
+ */
+function calculateFontSize(
+  measurer,
+  expectedHeight,
+  expectedWidth,
+  minFontSize,
+  maxFontSize
+) {
+  maxFontSize++;
+  // Binomial search for the best font size.
+  while (maxFontSize - minFontSize > 1) {
+    const mid = Math.floor((minFontSize + maxFontSize) / 2);
+    setStyle(measurer, 'fontSize', px(mid));
+    const width = measurer./*OK*/ scrollWidth;
+    const height = measurer./*OK*/ scrollHeight;
+    if (height > expectedHeight || width > expectedWidth) {
+      maxFontSize = mid;
+    } else {
+      minFontSize = mid;
+    }
+  }
+  setStyle(measurer, 'fontSize', px(minFontSize));
+  return minFontSize;
+}
+
+/**
+ * @param {Element} measurer
+ * @param {number} maxHeight
+ * @param {number} fontSize
+ */
+function getOverflowStyle(measurer, maxHeight, fontSize) {
+  const overflown = measurer./*OK*/ scrollHeight > maxHeight;
+  const lineHeight = fontSize * LINE_HEIGHT_EM_;
+  const numberOfLines = Math.floor(maxHeight / lineHeight);
+  if (overflown) {
+    setStyles(measurer, {
+      'lineClamp': numberOfLines,
+      '-webkit-line-clamp': numberOfLines,
+      'maxHeight': px(lineHeight * numberOfLines),
```

Comment:
So, this is one part I don't fully understand. I think this might be a bug for us, but not sure. If we set `maxHeight` and then the content actually needs more space we will never know, right? Because we won't get a resize event for a case with `neededHeight > maxHeight`.

Reply:
The `maxHeight` is in place when there is overflow content even with the minimum font size. If the content needs more space it would a require a change to `minFontSize`, which would cause a font size recalculation.

Reply:
But do we remove `maxHeight` before recalculating? If we don't, it'd seem that we'd never be able to expand beyond it.

Reply:
Good catch, I reproduced this by modifying the `min-font-size` on the overflow example to find the line clamp was still here. In the `else` block I replaced the `setStyles(measurer, {})` with `resetStyles` with the overflow CSS properties.